### PR TITLE
Fix various bugs

### PR DESCRIPTION
**Pull request by ghost**
- ~~full support for icv.db type pfs with file tables with more than one pages of signatures (fixes https://github.com/motoharu-gosuto/psvpfstools/issues/21 https://github.com/motoharu-gosuto/psvpfstools/issues/39 https://github.com/motoharu-gosuto/psvpfstools/issues/40 https://github.com/motoharu-gosuto/psvpfstools/issues/42)~~ (see PR https://github.com/motoharu-gosuto/psvpfstools/pull/47)
- fix wrong destination path (https://github.com/motoharu-gosuto/psvpfstools/issues/27)
- fix sce_sys wrongly excluded (https://github.com/motoharu-gosuto/psvpfstools/issues/22)
- fix padding not being trimmed in decrypted files from icv.db type pfs

### CODE REVIEW

**Comment by motoharu-gosuto** [comment]
@cuevavirus

Thank you for hard work! I would need some time to go through fixes. It is great, how much were you able to fix. It is also good that you have merged into io-api branch.

**Comment by motoharu-gosuto** [comment]
@cuevavirus
Commit: 5c218fbc0cfe868116ba34da79210fdf29e4e3a3

cmake/scripts/FindLIBTOMCRYPT.cmake
You have only left find_path, which I believe will not work on Windows. Unless CMAKE_PATH is provided. Can this change be reverted?
File itself was renamed from FindLibTomCrypt.cmake to FindLIBTOMCRYPT.cmake 
Is this change really required?

File: psvpfsparser/F00DNativeKeyEncryptor.cpp [hunk 1]
```diff
@@ -52,21 +52,21 @@ int F00DNativeKeyEncryptor::encrypt_key(const unsigned char* key, int key_size,
...
    if(kit != m_keyCache.end())
    {
-      std::uint32_t nbytes = key_size / 8;
       string_to_byte_array(kit->second, nbytes, drv_key);
       return 0;
    }
    else
    {
-      if(kprx_auth_service_0x50001(key, key_size, drv_key, 0) < 0)
+      if(kprx_auth_service_0x50001(key, nbytes, drv_key, 0) < 0)
```
Comment (motoharu-gosuto) [inline psvpfsparser/F00DNativeKeyEncryptor.cpp:66]: Good fix for key_size. I did not have time to test latest changes in io-api branch.

File: psvpfsparser/psvpfsparser.cpp [hunk 2]
```diff
@@ -77,10 +77,12 @@ std::shared_ptr<IF00DKeyEncryptor> create_F00D_encryptor(const PsvPfsParserConfi
    {
       case F00DEncryptorTypes::file:
          iF00D = F00DKeyEncryptorFactory::create(cfg.f00d_enc_type, cfg.f00d_arg); 
+         break;
```
Comment (motoharu-gosuto) [inline psvpfsparser/psvpfsparser.cpp:80]: Again. Good fix. I did not have time to test. Was planning to do further refactoring.

File: psvpfsparser/PsvPfsParserConfig.cpp [hunk 3]
```diff
@@ -100,11 +100,6 @@ int parse_options(int argc, char* argv[], PsvPfsParserConfig& cfg)
          f00d_cache = vm[F00D_CACHE_NAME].as<std::string>();
       }
 
-      if(f00d_url.empty() && f00d_cache.empty())
```
Comment (motoharu-gosuto) [inline psvpfsparser/PsvPfsParserConfig.cpp:103]: Can you explain this change. Why did you remove this check?

File: cmake/scripts/FindLIBTOMCRYPT.cmake [hunk 4]
```diff
@@ -7,19 +7,11 @@
 # LIBTOMCRYPT_INCLUDE_DIRS - The LIBTOMCRYPT include directories
 # LIBTOMCRYPT_LIBRARIES - The libraries needed to use LIBTOMCRYPT
 
-if(${LIBTOMCRYPT_INCLUDE_DIR})
```
Comment (motoharu-gosuto) [inline cmake/scripts/FindLIBTOMCRYPT.cmake:10]: You have only left find_path, which I believe will not work on Windows. Unless CMAKE_PATH is provided. Can this change be reverted?
File itself was renamed from FindLibTomCrypt.cmake to FindLIBTOMCRYPT.cmake
Is this change really required?

File: psvpfsparser/PsvPfsParserConfig.cpp [hunk 5]
```diff
@@ -108,9 +108,6 @@ int parse_options(int argc, char* argv[], PsvPfsParserConfig& cfg)
 
       if(!f00d_url.empty())
```
Comment (motoharu-gosuto) [inline psvpfsparser/PsvPfsParserConfig.cpp:109]: I agree with this commit. It is good that you did not remove url option.
I have made them hidden, but they are left for backwards compatibility.
Curl related code could be indeed removed.

File: psvpfsparser/Utils.cpp [hunk 6]
```diff
@@ -65,35 +65,28 @@ void getFileListNoPfs(boost::filesystem::path root_path, std::set<boost::filesys
...
-         if(boost::ends_with(cp, boost::filesystem::path("sce_pfs")))
-            continue;
-
-         //skip pfs inside sce_pfs (for ADDCONT)
-         if(boost::ends_with(cp, boost::filesystem::path("sce_pfs") / boost::filesystem::path("files.db")))
-            continue;
-
-         //skip pfs inside sce_pfs (for ADDCONT)
-         if(boost::ends_with(cp, boost::filesystem::path("sce_pfs") / boost::filesystem::path("unicv.db")))
+         if(cp == (root_path / boost::filesystem::path("sce_sys/package"))) {
```
Comment (motoharu-gosuto) [inline psvpfsparser/Utils.cpp]: can we use / operator instead of / in string constant?

File: psvpfsparser/Utils.cpp [hunk 7]
```diff
@@ -65,35 +65,28 @@ void getFileListNoPfs(boost::filesystem::path root_path, std::set<boost::filesys
...
+         }
 
          //skip pfs inside sce_sys (for ADDCONT)
-         if(boost::ends_with(cp, boost::filesystem::path("sce_sys")))
-         {
-            if(cp != root_path / boost::filesystem::path("sce_sys"))
-               continue;
+         if(boost::ends_with(cp, boost::filesystem::path("sce_sys")) &&
+               cp != root_path / boost::filesystem::path("sce_sys") &&
+               boost::filesystem::exists(cp / boost::filesystem::path("keystone"))) {
```
Comment (motoharu-gosuto) [inline psvpfsparser/Utils.cpp:85]: Can you explain why do we need to check keystone existence?

File: psvpfsparser/Utils.cpp [hunk 8]
```diff
@@ -96,6 +96,11 @@ void getFileListNoPfs(boost::filesystem::path root_path, std::set<boost::filesys
    }
 }
 
+boost::filesystem::path source_path_to_dest_path(const boost::filesystem::path& source_root, const boost::filesystem::path& dest_root, const boost::filesystem::path& source_path) {
+   boost::filesystem::path dest_path = dest_root / boost::filesystem::relative(source_path, source_root);
+   return boost::filesystem::path(dest_path.generic_string());
+}
+
```
Comment (motoharu-gosuto) [inline psvpfsparser/Utils.cpp:103]: Good fix! replace_all was doing a lot of trouble

File: psvpfsparser/FilesDbParser.cpp [hunk 9]
```diff
@@ -265,12 +265,6 @@ bool FilesDbParser::validate_header(uint32_t dataSize)
       return false;
    }
 
-   if(m_header.unk6 != 0xFFFFFFFFFFFFFFFF)
```
Comment (motoharu-gosuto) [inline psvpfsparser/FilesDbParser.cpp:268]: Please - do not remove this check.
If it can contain other values, as you have indicated in the comments - include those values in the check.
This way we can build up statistics on possible values and maybe understand what this field means.

File: psvpfsparser/FilesDbParser.cpp [hunk 10]
```diff
@@ -348,6 +342,14 @@ bool FilesDbParser::parseFilesDb(std::ifstream& inputStream, std::vector<sce_ng_
          return false;
       }
 
+      //bad block with error or unknown format
+      bool is_bad_block = false;
+      if (block.header.nFiles > 9)
```
Comment (motoharu-gosuto) [inline psvpfsparser/FilesDbParser.cpp]: So the error was caused by nFiles being greater than 9 ?
Please use the corresponding constant MAX_FILES_IN_BLOCK for this check.
Can you document title ids that have this issue? Just add the comment to the code.
This is a weird issue since this value should not be greater than 9.

File: psvpfsparser/FilesDbParser.cpp [hunk 11]
```diff
@@ -364,7 +366,7 @@ bool FilesDbParser::parseFilesDb(std::ifstream& inputStream, std::vector<sce_ng_
          std::vector<std::uint8_t> unusedData1(nUnusedSize1);
          inputStream.read((char*)unusedData1.data(), nUnusedSize1);
 
-         if(!isZeroVector(unusedData1))
+         if(!isZeroVector(unusedData1) && !is_bad_block)
```
Comment (motoharu-gosuto) [inline psvpfsparser/FilesDbParser.cpp]: Please add another check for is_bad_block == true.
Print a warning so that we can keep track of the errors in console.
I understand that this code is implemented in such a way that it skips invalid data, but user needs at least to know about it.

File: psvpfsparser/FilesDbParser.cpp [hunk 12]
```diff
@@ -348,6 +342,14 @@ bool FilesDbParser::parseFilesDb(std::ifstream& inputStream, std::vector<sce_ng_
          return false;
       }
 
+      //bad block with error or unknown format
+      bool is_bad_block = false;
+      if (block.header.nFiles > 9)
+      {
+         is_bad_block = true;
+         block.header.nFiles = 0;
```
Comment (motoharu-gosuto) [inline psvpfsparser/FilesDbParser.cpp:358]: Here you set nFiles to 0. I am assuming by doing that you are treating file records in this block as unused data. Is that correct? They will be read with nUnusedSize1.
Are you sure that by doing so, we can safely read file information records and hash table that goes after this data?
Maybe it is better to skip this block entirely?

File: psvpfsparser/PfsFile.cpp [hunk 13]
```diff
@@ -95,6 +95,10 @@ int PfsFile::decrypt_icv_file(boost::filesystem::path destination_root) const
    std::uintmax_t bytes_left = m_filepath.file_size();
    std::uint32_t sector_base = 0;
 
+   // icv.db pfs files are padded to the nearest sector boundary
+   // so we need to get the real size from files.db
+   std::uint32_t real_bytes_left = m_file.file.m_info.header.size;
```
Comment (motoharu-gosuto) [inline psvpfsparser/PfsFile.cpp]: i suggest renaming of real_bytes_left to page_size

File: psvpfsparser/PfsFile.cpp [hunk 14]
```diff
@@ -125,11 +129,20 @@ int PfsFile::decrypt_icv_file(boost::filesystem::path destination_root) const
...
+         if (read_size <= real_bytes_left)
+         {
+            outputStream.write((char*)buffer.data(), read_size);
+            real_bytes_left -= read_size;
+         }
+         else
+         {
+            outputStream.write((char*)buffer.data(), real_bytes_left);
+            real_bytes_left = 0;
+            break;
```
Comment (motoharu-gosuto) [inline psvpfsparser/PfsFile.cpp]: I am not sure if this is a good idea to break here.
If real_bytes_left is 0 and for cycle with blocks did not exit - it means something went wrong.
It is better to throw and error (return -1 and write an error to m_output) before trying to do a decryption if real_bytes_left != 0

File: psvpfsparser/PfsFilesystem.cpp [hunk 15]
```diff
@@ -122,7 +122,7 @@ int PfsFilesystem::decrypt_files(boost::filesystem::path destTitleIdPath) const
       //copy unencrypted files
       else if(is_unencrypted(file->file.m_info.header.type))
       {
-         if(!filepath.copy_existing_file(m_titleIdPath, destTitleIdPath))
+         if(!filepath.copy_existing_file(m_titleIdPath, destTitleIdPath, file->file.m_info.header.size))
```
Comment (motoharu-gosuto) [inline psvpfsparser/PfsFilesystem.cpp:125]: this fix should apply both to icv and unicv?

**Comment by motoharu-gosuto** [comment]
Also - I can not approve c09f2935f12e5b45ac992af6975fb932077dbccc right now.
It contains lots of core changes, which will require some time for me to remember and understand.

File: psvpfsparser/PsvPfsParserConfig.cpp (hunk 3 above)
Comment (ghost) [reply psvpfsparser/PsvPfsParserConfig.cpp:103]: If both url and cache options are not specified, then F00DNativeKeyEncryptor will be used. So the program should not terminate here.

File: cmake/scripts/FindLIBTOMCRYPT.cmake (hunk 4 above)
Comment (ghost) [reply cmake/scripts/FindLIBTOMCRYPT.cmake:10]: With the original code, cmake was unable to find libtomcrypt unless cmake was invoked a second time without deleting the build directory. I have changed it so that it builds for me on the first try and I think preserves what the original code was trying to do.

https://github.com/motoharu-gosuto/psvpfstools/blob/94471aefbc4e991bc2981f1489237df9c7ff44f2/cmake/scripts/FindLIBTOMCRYPT.cmake#L10-L22

The rename was to match case in ConfigureDependencies.cmake for case-sensitive environments. Alternatively, I could revert the file rename and change the names in ConfigureDependencies.cmake instead.

File: psvpfsparser/Utils.cpp (hunk 6 above)
Comment (ghost) [reply psvpfsparser/Utils.cpp]: fixed

https://github.com/motoharu-gosuto/psvpfstools/blob/94471aefbc4e991bc2981f1489237df9c7ff44f2/psvpfsparser/Utils.cpp#L77

File: psvpfsparser/Utils.cpp (hunk 7 above)
Comment (ghost) [reply psvpfsparser/Utils.cpp:85]: I believe sce_sys directories which exist in the same directory as a sce_pfs directory will always contain a keystone file.

This particular check is for skipping sce_sys directories of the form `<TITLE ID>/<ADDCONT ID>/sce_sys` when decrypting an addcont root. If the keystone existence check is not there, then this would also skip for example `PCSG01136/savedata/sce_sys/param.sfo`.

Possibly a better approach that could be implemented in the future is to use files.db to determine which files are part of the pfs instead of filtering by names and paths.

File: psvpfsparser/FilesDbParser.cpp (hunk 9 above)
Comment (ghost) [reply psvpfsparser/FilesDbParser.cpp:268]: added this back

https://github.com/motoharu-gosuto/psvpfstools/blob/94471aefbc4e991bc2981f1489237df9c7ff44f2/psvpfsparser/FilesDbParser.cpp#L268-L272

File: psvpfsparser/FilesDbParser.cpp (hunk 10 above)
Comment (ghost) [reply psvpfsparser/FilesDbParser.cpp]: added

https://github.com/motoharu-gosuto/psvpfstools/blob/94471aefbc4e991bc2981f1489237df9c7ff44f2/psvpfsparser/FilesDbParser.cpp#L351-L359

File: psvpfsparser/FilesDbParser.cpp (hunk 11 above)
Comment (ghost) [reply psvpfsparser/FilesDbParser.cpp]: added separate check and warning

https://github.com/motoharu-gosuto/psvpfstools/blob/94471aefbc4e991bc2981f1489237df9c7ff44f2/psvpfsparser/FilesDbParser.cpp#L377-L381

File: psvpfsparser/FilesDbParser.cpp (hunk 12 above)
Comment (ghost) [reply psvpfsparser/FilesDbParser.cpp:358]: I have seen only one case of this happening. With the changes I made, it decrypted successfully and matched exactly the decrypted data I got using vitashell.

[filesdb.zip](https://github.com/motoharu-gosuto/psvpfstools/files/4084773/filesdb.zip)

File: psvpfsparser/PfsFile.cpp (hunk 14 above)
Comment (ghost) [reply psvpfsparser/PfsFile.cpp]: changed

https://github.com/motoharu-gosuto/psvpfstools/blob/94471aefbc4e991bc2981f1489237df9c7ff44f2/psvpfsparser/PfsFile.cpp#L132-L146

https://github.com/motoharu-gosuto/psvpfstools/blob/94471aefbc4e991bc2981f1489237df9c7ff44f2/psvpfsparser/PfsFile.cpp#L158-L162

File: psvpfsparser/PfsFilesystem.cpp (hunk 15 above)
Comment (ghost) [reply psvpfsparser/PfsFilesystem.cpp:125]: As far as I can tell, unicv pfs files will always have the exact size instead of being padded.

File: psvpfsparser/PsvPfsParserConfig.cpp (hunk 3 above)
Comment (motoharu-gosuto) [reply psvpfsparser/PsvPfsParserConfig.cpp:103]: I see. That is correct.

File: cmake/scripts/FindLIBTOMCRYPT.cmake (hunk 4 above)
Comment (motoharu-gosuto) [reply cmake/scripts/FindLIBTOMCRYPT.cmake:10]: Everything is perfect now. Find script was indeed a mess.

File: psvpfsparser/Utils.cpp (hunk 7 above)
Comment (motoharu-gosuto) [reply psvpfsparser/Utils.cpp:85]: Yes. Keystone file is always inside sce_pfs directory and is always checked after decryption.
This check is done for both icv and unicv.

About usage of files.db. What you suggest is actually already implemented. However this `getFileListNoPfs` method is also required. Everything is a bit more complex. When parser finishes parsing files.db it validates the number of files.
Basically this helps us to verify that there are no extra files present in real filesystem which are not present in files.db. And the other way around as well. Parser intersects the list of real files vs list of files from files.db. However it excludes (masks) files that exist in files.db and not in real filesystem (nonexistent files can be specified in files.db). It also excludes (masks) files that exist in real filesystem but are expected not to be present in files.db. For example contents of sce_pfs directory. You can see this logic in the end of `int FilesDbParser::parse()` method.

File: psvpfsparser/Utils.cpp (hunk 6 above)
Comment (motoharu-gosuto) [reply psvpfsparser/Utils.cpp]: perfect.

File: psvpfsparser/FilesDbParser.cpp (hunk 9 above)
Comment (motoharu-gosuto) [reply psvpfsparser/FilesDbParser.cpp:268]: Perfect. This is very interesting that unk6 equals 0x400. Which is EXPECTED_BLOCK_SIZE in files.db. I wonder if this is a coincidence.

File: psvpfsparser/FilesDbParser.cpp (hunk 10 above)
Comment (motoharu-gosuto) [reply psvpfsparser/FilesDbParser.cpp]: Perfect. Both usage of constant and provided information.

File: psvpfsparser/FilesDbParser.cpp (hunk 12 above)
Comment (motoharu-gosuto) [reply psvpfsparser/FilesDbParser.cpp:358]: Thanks for the provided information. This will require further investigation.
Please create a new clean issue and attach the file there.
We need to formulate the symptoms in the issue.
It looks like this unk6 == 0x400 and nFiles > 9 (which probably was -1 ?)
is some weird condition for block skipping. Just my guess.

File: psvpfsparser/FilesDbParser.cpp (hunk 11 above)
Comment (motoharu-gosuto) [reply psvpfsparser/FilesDbParser.cpp]: perfect

File: psvpfsparser/PfsFile.cpp (hunk 14 above)
Comment (motoharu-gosuto) [reply psvpfsparser/PfsFile.cpp]: perfect

File: psvpfsparser/PfsFilesystem.cpp (hunk 15 above)
Comment (motoharu-gosuto) [reply psvpfsparser/PfsFilesystem.cpp:125]: Ok. I was just curious if copy_existing_file overload could cause some side effects because decrypt_files method is called for both icv and unicv files. I guess not.

Comment (ghost) [reply psvpfsparser/PfsFilesystem.cpp:125]: sorry misunderstood what you meant earlier. But yeah I think it is fine like this.

**Comment by ghost** [comment]
I rebased 048edbfe out of this pull request. When this PR is merged I will open a separate PR for that.
